keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,974 @@
+
+Another problem is that log messages during program startup need to be handled
+somehow. We then can either:
+- discard the message, which would break the current `klog.Error(...);
+  os.Exit(1)` code pattern when an error occurs during initialization log
+- log with some default configuration (for example, as text)
+
+In both cases we have the problem that a logger retrieved by some init code
+before logging initialization will continue to use that behavior (discard or
+log with default configuration) even after logging gets initialized.
+
+Ideally, no log messages should be emitted before the program is done with
+setting up logging as intended. This is already [problematic
+now](https://github.com/kubernetes/kubernetes/issues/100152) because output may
+change from text (the current klog default) to JSON (once initialized). There
+will be no automatic mitigation for this. Such log calls will have to be found
+and fixed manually, for example by [passing the error back to `main`
+instead](https://github.com/kubernetes/kubernetes/pull/104774), which is part
+of an effort to [remove the dependency on logging before an unexpected program
+exit](https://github.com/kubernetes/kubernetes/issues/102231).
+
+To support identifying log messages that need to be fixed, the default logger
+during program startup will be using the klog text output instead of discarding
+the output and a `INIT-LOG-FIXME` prefix will be inserted into the log output.
+
+##### Performance overhead
+
+Retrieving a logger from a context on each log call will have a higher overhead
+than using a global logger instance. The overhead will be measured for
+different scenarios. If a function uses a logger repeatedly, it should retrieve
+the logger once and then use that instance.
+
+#### Transition
+
+Once some components start to rely on `k8s.io/klogr` exclusively instead of
+klog, users of those components must remember to initialize klog *and* the
+default logger in `k8s.io/klogr`. Features provided only by klog like writing
+to files become unavailable.
+
+This breaking change can be postponed until after the [deprecation of klog
+features](https://github.com/kubernetes/enhancements/issues/2845) is complete
+by using a logger which passes log messages through to klog. That allows
+converting components already now. Then later, the klog dependency in
+`k8s.io/klogr` can be replaced with a stand-alone logger which produces the same
+text output.
+
+#### Pitfalls during usage
+
+Code reviews must catch the following incorrect patterns.
+
+#### Logger in context and function not consistent
+
+Incorrect:
+
+```go
+func foo(ctx context.Context) {
+   logger := klogr.FromContext(ctx)
+   ctx = klogr.NewContext(ctx, logger.WithName(“foo”))
+   doSomething(ctx)
+   // BUG: does not include WithName(“foo”)
+   logger.Info(“Done”)
+}
+```
+
+Correct:
+
+```go
+func foo(ctx context.Context) {
+   logger := klogr.FromContext(ctx).WithName(“foo”)
+   ctx = klogr.NewContext(ctx, logger)
```

Comment:
do we expect nesting/overriding the logger to be common? every time that happens, it increases the cost of looking up any other value from the context

Reply:
I expect this to be used infrequently. The `WithName` part accumulates, so developers would see a very long prefix in the log output if it was used excessively.

The same applies to `WithValues`: developers need to make a conscious choice that some value is important and will see the effect.

Reply:
> Probably worth calling it out in the "performance overhead" section.

Done.

> Also, are you planning to extract some of these notes into documentation (perhaps in the community repo)? If so, worth mentioning in the graduation criteria.

We have https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md. That could be extended or replaced. Much of the necessary material (coding style, caveats) is already in the KEP. I've added it to the alpha graduation criteria.

Reply:
> Can this be caught with a linter?

No, I don't think so. A linter cannot know whether the code is performance sensitive, so under which conditions should it warn about this? It's perfectly fine to do this in code which cares more about informative log messages than performance.

Reply:
> I expect this to be used infrequently.

that doesn't seem to mesh with the guidance that callees only add local values when logging and rely on callers to create loggers with any values they want logged ... xref https://github.com/kubernetes/enhancements/pull/3078/files#r797717281

Reply:
links don't help when github hides the file by default :(

If ther are a lot of discrete calls to `WithValues` it feels like someone is doing something wrong.  E.g. a linter COULD catch:

```
log = log.WithValues("foo", foo)
...
log = log.WithValues("bar", bar)
```

which can flatten to a single call and half as many allocations.

Reply:
and a linter could catch:

```
ctx = klogr.NewContext(ctx, logger.WithName("foo"))
```

and suggest breaking it to 2 statements

Reply:
As mentioned in the other conversation, this guidance here is correct and the section about potentially duplicating key/value pairs was updated.

As you said "could" I'll take these as suggestions, but not add it as "required" to the KEP. I'll probably end implementing it soon anyway, hacking on logcheck has given me a pretty good understanding of how linters work...